pkg/injector/bootstrap.go

```diff
@@ -100,3 +105,103 @@ func (wh *mutatingWebhook) marshalAndSaveBootstrap(name, namespace string, confi
 	log.Debug().Msgf("Creating bootstrap config for Envoy: name=%s, namespace=%s", name, namespace)
 	return wh.kubeClient.CoreV1().Secrets(namespace).Create(context.Background(), secret, metav1.CreateOptions{})
 }
+
+// NewBootstrapSecretRotator returns a new bootstrap secret rotator.
+func NewBootstrapSecretRotator(ctx context.Context, kubeClient kubernetes.Interface, informerCollection *informers.InformerCollection, certManager *certificate.Manager, checkInterval time.Duration) *BootstrapSecretRotator {
+	return &BootstrapSecretRotator{
+		context:            ctx,
+		kubeClient:         kubeClient,
+		informerCollection: informerCollection,
+		certManager:        certManager,
+		checkInterval:      checkInterval,
+	}
+}
+
+// listBootstrapSecrets returns the bootstrap secrets stored in the informerCollection's store.
+// this function is used to get the namespace of the secrets.
+func (b *BootstrapSecretRotator) listBootstrapSecrets() []*corev1.Secret {
+	// informers return slice of pointers so we'll convert them to value types before returning
+	secretPtrs := b.informerCollection.List(informers.InformerKeySecret)
+	var secrets []*corev1.Secret
+
+	for _, secretPtr := range secretPtrs {
+		if secretPtr == nil {
+			continue
+		}
+		secret, ok := secretPtr.(*corev1.Secret)
+		if !ok {
+			continue
+		}
+		// finds bootstrap secrets
+		if strings.Contains(secret.Name, bootstrapSecretPrefix) {
+			secrets = append(secrets, secret)
+		}
+	}
+
+	return secrets
+}
+
+// rotateBootstrapSecrets finds the bootstrap secret of the given certificate
+// from the list of secrets stored in the informerCollection's store and updates the secret.
+func (b *BootstrapSecretRotator) rotateBootstrapSecrets() {
+	secrets := b.listBootstrapSecrets()
+	for _, secret := range secrets {
+		secretProxyUUID := strings.ReplaceAll(secret.Name, bootstrapSecretPrefix, "")
+		certs := b.certManager.ListIssuedCertificates()
+		for _, cert := range certs {
+			// CommonName for a xds certificate has the form: <ProxyUUID>.<kind>.<identity>
+			certProxyUUID := strings.Split(cert.CommonName.String(), ".")[0]
+			// checks if this is the corresponding cert for the secret
+			if secretProxyUUID != certProxyUUID {
+				continue
+			}
+			opts := []certificate.IssueOption{certificate.FullCNProvided()}
+			c, err := b.certManager.IssueCertificate(cert.CommonName.String(), cert.CertType, opts...)
+			if err != nil {
+				log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrRotatingCert)).
+					Msgf("Error rotating cert SerialNumber=%s", cert.GetSerialNumber())
+			}
+			// if the secret and issued cert are the same no need to update the secret
+			if reflect.DeepEqual(c.GetIssuingCA, secret.Data["ca.crt"]) && reflect.DeepEqual(c.PrivateKey, secret.Data["tls.key"]) && reflect.DeepEqual(c.CertChain, secret.Data["tls.crt"]) {
+				continue
+			}
+			// TODO: check mapping
+			secretData := map[string][]byte{
+				"ca.crt":  cert.GetIssuingCA(),
```

Comment:
```suggestion
				"ca.crt":  cert.GetTrustedCA(),
```
The issuing ca is the ca that was used to sign the issued certificate. In this case, we want to set the ca.crt to the trusted cas of the certificate. The trusted ca will contain the issuing ca and may additionally contain the validating ca during root certificate rotation.

During some stages of rotation, all certificates must have both cas available for validation. For example, there may be a set of certs issued by one ca and a set of certs issued by another. In order for those services (or in this case the ADS server and envoy) to communicate with mTLS both services must trust both cas.